### PR TITLE
Add a new flag for ingestion to preserve existing metrics

### PR DESCRIPTION
**Pull request by maytasm**
Add a new flag for ingestion to preserve existing metrics

### Description

This PR adds a new flag in the tuningSpec as part of the AppendableIndexSpec. When set to true, ingestion will do the following for each given metric in the metricSpec depending on the existing state of the segment:
- The segment does not have the metric name defined in metricSpec as any of it’s metric → the metric aggregator in metricSpec is applied if the source column exist otherwise the metric value is null
- The segment has the metric name defined in metricSpec as one of it’s metric → the metric aggregator in metricSpec is skipped and the existing metric is unchanged

User shouldn’t need to set this directly and will automatically be set by the auto compaction.

Note on some edge cases / sad path:
- If non-rolled up rows has columns with the same name as aggregators in metric spec (after rolling up) and isPreserveExistingMetrics is enabled, then those non-rolled up columns would be treated as metrics and preserved (we wont compute the metric again from the source column for those rows).
- Compaction task will fail if isPreserveExistingMetrics is enabled and there are rows with conflicting types in the same segment that cannot be aggregated. For example, we have String and Number in column foo, or Number in column sum_foo and String in column foo....and we have LongSum Aggregator with fieldName=foo and output=sum_foo. 
- Compaction task will still succeed when isPreserveExistingMetrics is enabled and there are rows with conflicting types across different segments that cannot be aggregated. For example, we have String in column foo of segment A and Number in column foo of segment B....and we have LongSum Aggregator with fieldName=foo and output=sum_foo. The combining agg will ignore the incorrect type and ingestion would succeed. This is also the same as current behavior without the preserve existing metrics flag. This PR (and the flag isPreserveExistingMetrics) does not modify anything regarding combining metrics across intermediate segments. Hence, this is the same behavior as current ingestion without the isPreserveExistingMetrics.

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java [hunk 1]
```diff
@@ -725,4 +728,171 @@ public void testDynamicSchemaRollup() throws IndexSizeExceededException
 
     Assert.assertEquals(2, index.size());
   }
+
```
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java:731]: do we need sad path tests here? like is it possible to end up with a mix of aggregators that are not the same type if the ingestion schema changed between segment creation? what if non-rolled up has columns with the same name as aggregators in rolled up? should these be an error? is there some external safeguards that prevent these tasks from ever getting started due to incompatible segment schemas to use this mode that is external to this PR?

also what about more extreme cases, like sketch aggregators, which have a lot more variety between their building aggregators and combining aggregators?

basically my concern here is around making sure there are no scenarios where we are silently doing the wrong thing, writing incorrect segments, instead of loudly failing when it isn't well supported.

File: processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java [hunk 2]
```diff
@@ -258,7 +259,8 @@ public ColumnCapabilities getColumnCapabilities(String columnName)
   protected IncrementalIndex(
       final IncrementalIndexSchema incrementalIndexSchema,
       final boolean deserializeComplexMetrics,
-      final boolean concurrentEventAdd
+      final boolean concurrentEventAdd,
+      final boolean preserveExistingMetrics
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:263]: i feel like this definitely needs javadocs/comments to ensure that nothing ever sets this except DruidInputSource, since that is the only case where we should have existing metrics

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 3]
```diff
@@ -89,10 +93,11 @@
       boolean concurrentEventAdd,
       boolean sortFacts,
       int maxRowCount,
-      long maxBytesInMemory
+      long maxBytesInMemory,
+      boolean preserveExistingMetrics
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:97]: same comment about javadoc and comments about reingestion

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 4]
```diff
@@ -435,6 +490,18 @@ public boolean isNull(int rowOffset, int aggOffset)
     }
   }
 
+  private Object getCombinedMetric(AggregatorFactory[] metrics, Aggregator[] aggs, int aggOffset)
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:493]: this seems worthy of javadoc/comments

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 5]
```diff
@@ -157,6 +163,16 @@ protected void initAggs(
               concurrentEventAdd
           )
       );
+      if (isPreserveExistingMetrics()) {
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:166]: nit: since this is happening basically per column per row, `isPreserveExistingMetrics()` should probably be saved in a final field or just made `protected` in `IncrementalIndex`, and use that directly which probably has a better chance of being optimized by jvm than this being a method call everywhere (though maybe jvm is cool enough it never usually hurts to help it out by being explicit and cutting down method invocations). same applies to other uses of `isPreserveExistingMetrics()`

side note: it might be worth sanity check measuring performance of ingestion before and after this change to make sure there is little/no impact when this mode is on

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 6]
```diff
@@ -353,31 +383,51 @@ public String getOutOfRowsReason()
...
 
   @Override
   protected double getMetricDoubleValue(int rowOffset, int aggOffset)
   {
-    return concurrentGet(rowOffset)[aggOffset].getDouble();
+    if (isPreserveExistingMetrics()) {
+      return (double) getCombinedMetric(getMetricAggs(), concurrentGet(rowOffset), aggOffset);
+    } else {
+      return concurrentGet(rowOffset)[aggOffset].getDouble();
+    }
```
Comment (liran-funaro) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:420]: This `if/else` logic repeats itself. It makes sense to put it in a function or move the entire logic to `getCombinedMetric()`.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 7]
```diff
@@ -248,9 +268,14 @@ private void factorizeAggs(
   )
   {
     rowContainer.set(row);
-    for (int i = 0; i < metrics.length; i++) {
+    int aggLength = isPreserveExistingMetrics() ? aggs.length / 2 : aggs.length;
```
Comment (liran-funaro) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:271]: You can use here `metrics.length` instead.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 6 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:482]: Done

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 7 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:302]: Done

File: processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java (hunk 2 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java]: Done

File: processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java (hunk 1 above)
Comment (maytasm) [reply processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java:732]: - Mix of aggregators with incompatible type. (i.e. compaction with LongSum but some segment has existing metric as String and some has number) is a sad path and will result in the ingestion task failing. The error message will mentioned that this is because isPreserveExistingMetrics is enabled (isPreserveExistingMetrics can be disable to fallback to original behavior). Added this as a test case.
- If non-rolled up has columns with the same name as aggregators in rolled up and isPreserveExistingMetrics is enabled, then those non-rolled up columns would be treated as metrics and combined.
- There is currently no safeguards (other than that the job will fail without creating new unexpected segments). We can look into adding these safeguard later to make the job fail fast. 
- Also tested hll, theta, and quantilesDoublesSketch aggregators via the Integration Tests.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 3 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java]: Done

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 5 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:200]: Done

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 4 above)
Comment (maytasm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java]: This method is renamed to getMetricHelper and does have javadoc now.

**Review by clintropolis — APPROVED** [review]

File: core/src/main/java/org/apache/druid/data/input/InputRowSchema.java [hunk 8]
```diff
@@ -31,15 +34,38 @@
...
+      @Nullable final List<String> metricNames
+  )
   {
     this.timestampSpec = timestampSpec;
     this.dimensionsSpec = dimensionsSpec;
     this.columnsFilter = columnsFilter;
+    this.metricNames = metricNames;
+  }
+
+  @Nullable
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/InputRowSchema.java:65]: How should callers of this method think about the difference between a null list and an empty list?

It looks like the code behaves the same right now whether or not the list is null or empty. Can we update this to be `@NotNull` and return an empty list if metricNames is null. This will make it easier for callers of this method as they do not have to add null checks.

If null is meant to represent a different state, can we please add javadocs to explain that difference.

File: core/src/main/java/org/apache/druid/data/input/InputRowSchema.java [hunk 9]
```diff
@@ -31,15 +34,38 @@
   private final DimensionsSpec dimensionsSpec;
   private final ColumnsFilter columnsFilter;
 
+  /**
+   * List of metric names for further downstream processing by {@link InputSource}
+   */
+  @Nullable
+  private final List<String> metricNames;
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/InputRowSchema.java:41]: Should this be a set, or does the order of metricNames matter?

**Review by suneet-s — COMMENTED** [review]
Overall, this makes sense to me. The position of the feature flag, being part of the AppendableIndexBuilder seems like the best solution of lots of bad options. Ideally, I think this should have been part of metricsSpec, but metricsSpec is an array of AggregatorFactories, which means introducing this config there could be painful.

Let's mark this config as unstable so other developers know not to use it / rely on it. Only the auto-compaction system can use it behind the scenes for now, and if / when we think of a better solution, we can come up with a migration plan.

File: integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java [hunk 10]
```diff
@@ -110,6 +118,226 @@ public void setup() throws Exception
...
+          "%%EXPECTED_SCAN_RESULT%%", ImmutableList.of(ImmutableMap.of("events", ImmutableList.of(nullList)), ImmutableMap.of("events", ImmutableList.of(ImmutableList.of(31))))
+      );
+      verifyQuery(INDEX_ROLLUP_QUERIES_RESOURCE, expectedResult);
+      expectedResult = ImmutableMap.of(
+          "%%FIELD_TO_QUERY%%", "count",
+          "%%EXPECTED_COUNT_RESULT%%", 2,
+          "%%EXPECTED_SCAN_RESULT%%", ImmutableList.of(ImmutableMap.of("events", ImmutableList.of(ImmutableList.of(2))), ImmutableMap.of("events", ImmutableList.of(nullList)))
+      );
+      verifyQuery(INDEX_ROLLUP_QUERIES_RESOURCE, expectedResult);
+      expectedResult = ImmutableMap.of(
```
Comment (suneet-s) [inline integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:147]:
```suggestion
      queryAndResultFields = ImmutableMap.of(
```

Naming is hard ¯\_(ツ)_/¯

File: core/src/main/java/org/apache/druid/data/input/InputRowSchema.java (hunk 8 above)
Comment (maytasm) [reply core/src/main/java/org/apache/druid/data/input/InputRowSchema.java:65]: There is no difference. Modified to make this clear.

File: core/src/main/java/org/apache/druid/data/input/InputRowSchema.java (hunk 9 above)
Comment (maytasm) [reply core/src/main/java/org/apache/druid/data/input/InputRowSchema.java:41]: It can be a set. The order does not matter

File: integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java [hunk 11]
```diff
@@ -110,6 +118,226 @@ public void setup() throws Exception
...
+          "%%FIELD_TO_QUERY%%", "sum_added",
+          "%%EXPECTED_COUNT_RESULT%%", 2,
+          "%%EXPECTED_SCAN_RESULT%%", ImmutableList.of(ImmutableMap.of("events", ImmutableList.of(ImmutableList.of(62))), ImmutableMap.of("events", ImmutableList.of(nullList)))
+      );
+      verifyQuery(INDEX_ROLLUP_QUERIES_RESOURCE, expectedResult);
```
[the remark this answers is not in the document]
Comment (maytasm) [reply integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:152]: Renamed

File: integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java (hunk 10 above)
Comment (maytasm) [reply integration-tests/src/test/java/org/apache/druid/tests/coordinator/duty/ITAutoCompactionTest.java:147]: Done

**Review by suneet-s — APPROVED** [review]
+1 after CI
